node/network/collator-protocol/src/validator_side/peer_slots.rs

```diff
@@ -0,0 +1,1046 @@
+// purposes of determining a collator's fitness. Metrics for collators that do not fit in the
+// Cache are stored in a database.
+const CACHE_SIZE: usize = 1_000;
+// RESERVOIR SIZE determines the upper bound on the number of collator connections
+#[cfg(not(test))]
+pub(super) const RESERVOIR_SIZE: usize = 128;
+#[cfg(test)]
+pub(super) const RESERVOIR_SIZE: usize = 16;
+
+/// Message was went out-of-order
+const COST_UNEXPECTED_MESSAGE: Rep = Rep::CostMinor("An unexpected message");
+/// Message could not be decoded properly.
+const COST_CORRUPTED_MESSAGE: Rep = Rep::CostMinor("Message was corrupt");
+/// Network errors that originated at the remote host should have same cost as timeout.
+const COST_NETWORK_ERROR: Rep = Rep::CostMinor("Some network error");
+pub(super) const COST_REQUEST_TIMED_OUT: Rep =
+	Rep::CostMinor("A collation request has timed out");
+pub(super) const COST_INVALID_SIGNATURE: Rep =
+	Rep::Malicious("Invalid network message signature");
+const COST_WRONG_PARA: Rep =
+	Rep::Malicious("A collator provided a collation for the wrong para");
+pub(super) const COST_UNNEEDED_COLLATOR: Rep = Rep::CostMinor("An unneeded collator connected");
+
+pub(super) const COST_REPORT_BAD: Rep =
+	Rep::Malicious("A collator was reported by another subsystem");
+pub(super) const BENEFIT_NOTIFY_GOOD: Rep =
+	Rep::BenefitMinor("A collator was noted good by another subsystem");
+
+pub(super) const NUM_METRICS: usize = 7;
+pub(super) const COLLATOR_METRICS: [Rep; NUM_METRICS] = [
+	COST_UNEXPECTED_MESSAGE,
+	COST_CORRUPTED_MESSAGE,
+	COST_NETWORK_ERROR,
+	COST_REQUEST_TIMED_OUT,
+	COST_INVALID_SIGNATURE,
+	COST_WRONG_PARA,
+	COST_UNNEEDED_COLLATOR,
+];
+
+#[derive(Eq, PartialEq, Copy, Clone)]
+#[repr(usize)]
+pub enum FitnessEvent {
+	Unexpected = 0,
+	Corrupted = 1,
+	NetworkError = 2,
+	Timeout = 3,
+	SigError = 4,
+	ParaError = 5,
+	Superfluous = 6,
+}
+
+#[derive(Copy, Clone)]
+pub struct FitnessMetric {
+	last: std::time::Instant,
+	cumulative: Duration,
+	total: u64,
+}
+
+impl PartialEq for FitnessMetric {
+	fn eq(&self, other: &Self) -> bool {
+		(self.fitness() - other.fitness()).abs() < f64::EPSILON
+	}
+}
+
+impl Eq for FitnessMetric {}
+
+impl PartialOrd for FitnessMetric {
+	fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Ord for FitnessMetric {
+	fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+		if (self.fitness() - other.fitness()).abs() < f64::EPSILON {
+			std::cmp::Ordering::Less
+		} else {
+			std::cmp::Ordering::Greater
+		}
+	}
+}
+
+impl FitnessMetric {
+	pub fn new() -> Self {
+		Self {
+			last: std::time::Instant::now(),
+			cumulative: Duration::new(0, 0),
+			total: 0u64,
+		}
+	}
+
+	pub fn insert_event(&mut self, elapsed: Duration) {
+		self.total += 1;
+		self.cumulative += elapsed;
+		self.last = std::time::Instant::now();
+	}
+
+	pub fn elapsed(&self) -> Duration {
+		self.last.elapsed()
+	}
+
+	pub fn fitness(&self) -> f64 {
+		if self.total != 0 {
```

Comment:
I'd check here `self.cumulitive.as_nanos() != 0` just to be sure, in case elapsed event is zero for some reason.